index.bs

```diff
@@ -68,12 +71,145 @@ The appropriate times to <a spec=HTML>fetch and process the linked resource</a>
+  To <dfn export>parse speculation rules</dfn> given a [=string=] |input| and a [=URL=] |baseURL|, perform the following steps. They return a [=speculation rule set=] or null.
+
+  1. Let |parsed| be the result of [=Parse JSON into Infra values|parsing JSON into Infra values=] given |input|.
+  1. If |parsed| is not a [=map=], then return null.
+  1. Let |result| be an empty [=speculation rule set=].
+  1. If |parsed|["`prefetch`"] [=map/exists=] and is a [=list=], then [=list/for each=] |prefetchRule| of |parsed|["`prefetch`"]:
+    1. If |prefetchRule| is not a [=map=], then [=iteration/continue=].
+    1. Let |rule| be the result of [=parse a speculation rule|parsing a speculation rule=] given |prefetchRule| and |baseURL|.
+    1. If |rule| is null, then [=iteration/continue=].
+    1. [=list/Append=] |rule| to |result|'s [=speculation rule set/prefetch rules=].
+  1. Return |result|.
+</div>
+
+<div algorithm="parse a speculation rule">
+  To <dfn export>parse a speculation rule</dfn> given a [=map=] |input| and a [=URL=] |baseURL|, perform the following steps. They return a [=speculation rule=] or null.
+
+  1. If |input| has any [=map/key=] other than "`source`", "`urls`", and "`requires`", then return null.
+  1. If |input|["`source`"] does not [=map/exist=] or is not the [=string=] "`list`", then return null.
+  1. Let |urls| be an empty [=list=].
+  1. If |input|["`urls`"] does not [=map/exist=], is not a [=list=], or has any element which is not a [=string=], then return null.
+  1. [=list/For each=] |urlString| of |input|["`urls`"]:
+    1. Let |parsedURL| be the result of [=basic URL parser|parsing=] |urlString| with |baseURL|.
+    1. If |parsedURL| is failure, then [=iteration/continue=].
+    1. If |parsedURL|'s [=url/scheme=] is not an [=HTTP(S) scheme=], then [=iteration/continue=].
+    1. [=list/Append=] |parsedURL| to |urls|.
+  1. Let |requirements| be an empty [=ordered set=].
+  1. If |input|["`requires`"] [=map/exists=], but is not a [=list=], then return null.
+  1. [=list/For each=] |requirement| of |input|["`requires`"]:
+    1. If |requirement| is not the [=string=] "`anonymous-client-ip`", then return null.
+    1. [=set/Append=] |requirement| to |requirements|.
+  1. Return a [=speculation rule=] with [=speculation rule/URLs=] |urls| and [=speculation rule/requirements=] |requirements|.
+</div>
+
+<h3 id="speculation-rules-processing">Processing Model</h3>
+
+A [=document=] has a <dfn for=document export>list of speculation rule sets</dfn>, which is an initially empty [=list=].
+
+<!-- TODO(domfarolino): Get rid of the `data-link-type="interface"` once we fix the dfn in HTML. -->
+Periodically, the user agent may [=queue a task=] to the <a data-link-type="interface">DOM manipulation task source</a> to [=consider speculation=] for any [=document=].
+
+<p class="note">
+  The user agent will likely do this after the insertion of new speculation rules, or when resources are idle and available.
```

Comment:
I'm starting to wonder what the flow of this should look like. Should we directly invoke "consider speculation" from the steps you added where we mark a `speculationrules` script "ready"? It just feels weird that we mention below that UAs "likely" do this after the insertion of new speculation rules, when in fact we can actually make this happen if we want.

Reply:
Yeah, so my thinking here is that the rules determine what the UA could start, but I don't want to compel UAs to start prefetching immediately -- especially if in a future we allow rules that say things like "any link on this page". Instead I'm trying to express that when it's a good time (page is loaded, CPU is idle, network is fast, the user is hovering over a corresponding link right now, whatever) the UA may choose to do one or more permitted prefetches.

So I want to say that the UA won't do this in the middle of a task (so that you can make multiple changes without being surprised), but otherwise isn't confined to acting immediately. I'm reluctant to spec exactly what those heuristics are.

This sort relates to your other comments. In my view, rules should remain "in effect" and it should be up to the UA to not do silly things like repeatedly fetch the same URL. In fact maybe there are circumstances where it might want to -- if you walk away from your computer but then interact in a way that makes it likely you're going to navigate (like hovering a link), the first prefetch might be stale but a fresh one would be fruitful.

One of the concerns about prerender in particular is that it's hard to hold correctly without taking too much of a performance hit -- this lets the UA take a little more of the burden of deciding when it's fruitful to pre*.

Reply:
All of that makes sense, however we've gotten a good share of pushback when it comes to spec'ing things that have magical heuristics before. This might not be as bad since it is less visible on the referrer page etc, and solely an optimization, but we had a lot of push back when it came to automatic lazyload heuristics for example.